src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -252,34 +252,30 @@ public boolean isNative() {
     }
 
     @Override
-    public final MemorySegment overlap(MemorySegment other) {
+    public final MemorySegment asOverlappingSlice(MemorySegment other) {
         AbstractMemorySegmentImpl that = (AbstractMemorySegmentImpl)Objects.requireNonNull(other);
-        long offsetFromThis = this.offsetOf(that);
-        if (offsetFromThis >= 0L) {  // this and that overlap
-            long offsetFromThat = that.offsetOf(this);
-            long newSize = Math.min(this.byteSize() - offsetFromThis, that.byteSize() - offsetFromThat);
-            return dup(offsetFromThis, newSize, mask, scope);
-        }
-        return null;
-    }
-
-    @Override
-    public final long offsetOf(MemorySegment other) {
-        AbstractMemorySegmentImpl that = (AbstractMemorySegmentImpl) Objects.requireNonNull(other);
-        if (base() == that.base()) {  // both are either native or heap
+        if (base() == that.base()) {  // both either native or heap
             final long thisStart = this.min();
             final long thatStart = that.min();
             final long thisEnd = thisStart + this.byteSize() - 1L;
```

Comment:
do you need the `-1L` (since then you need `<=`) ? E.g. can we drop the `- 1L` and use strict comparisons instead?

Reply:
Yup, that's better!